### ISSUE TITLE
HAMi blocks NVLink P2P detection even with CUDA_DISABLE_CONTROL=true (libvgpu.so still intercepts P2P capability queries)

### ISSUE THREAD

**Issue by ok-claven** [Project-HAMi/HAMi#2760]
# HAMi Blocks NVLink P2P Detection Even With `CUDA_DISABLE_CONTROL=true`

## Problem

On a KVM VM with 3x NVIDIA H100 NVL GPUs where NVLink is active between GPU 0 and GPU 1 (12 links, ~318 GB/s aggregate), vLLM with `tensor-parallel-size=2` cannot use NVLink-accelerated communication. All NVLink-dependent optimizations fail with `CUDA driver error: invalid device ordinal`. NCCL falls back to SHM transport, and vLLM falls back to plain PYNCCL all-reduce — leaving significant performance on the table.

The deployment starts and serves requests, but every NVLink feature is silently disabled:

```
SymmMemCommunicator: symmetric memory initialization failed: CUDA driver error: invalid device ordinal
Custom allreduce is disabled because your platform lacks GPU P2P capability or P2P test failed
Failed to initialize FlashInfer All Reduce workspace: CUDA driver error: invalid device ordinal
FlashInfer mnnvl allreduce workspace unavailable (likely no NVSwitch multicast support); falling back to trtllm backend for single node.
Failed to initialize Flashinfer allreduce workspace. Flashinfer allreduce-norm fusion will be disabled.
AllReduce fusion pass is disabled.
```

Related: #2641 (that issue covers the NCCL init hang fixed by `CUDA_DISABLE_CONTROL=true`; this issue covers a *different* symptom — NVLink P2P detection is blocked even with that flag set).

## Environment

| Component | Details |
|-----------|---------|
| Node | KVM VM (internal host) |
| GPUs | 3x NVIDIA H100 NVL (H100L 94GB, sm_90) |
| NVLink | **Active** between GPU 0 ↔ GPU 1: 12 links at 26.562 GB/s each (~318 GB/s aggregate). GPU 2 has no NVLink. |
| Topology | `nvidia-smi topo -m`: GPU0↔GPU1 = `NV12`, GPU2 = `PHB` (PCIe only) |
| Driver | 595.84, CUDA 13.2 |
| vLLM | v0.26.0 (`vllm/vllm-openai:v0.26.0`), NCCL 2.28.9 |
| HAMi | v2.9.0, `hami-core` mode, `LD_PRELOAD=/usr/local/vgpu/libvgpu.so` |
| GPU Allocation | Whole-GPU mode (`nvidia.com/gpu: 2`, no `gpumem` fraction) |
| Env | `CUDA_DISABLE_CONTROL=true` set (required to fix the NCCL init hang — see #2641) |

### NVLink is active on the host VM

```
$ nvidia-smi nvlink -s
GPU 0: NVIDIA H100 NVL
  Link 0: 26.562 GB/s    Link 1: 26.562 GB/s    Link 2: 26.562 GB/s    Link 3: 26.562 GB/s
  Link 6: 26.562 GB/s    Link 7: 26.562 GB/s    Link 8: 26.562 GB/s    Link 9: 26.562 GB/s
  Link 12: 26.562 GB/s   Link 13: 26.562 GB/s   Link 14: 26.562 GB/s   Link 15: 26.562 GB/s

GPU 1: NVIDIA H100 NVL
  (same 12 active links)

$ nvidia-smi topo -m
       GPU0   GPU1   GPU2
GPU0    X    NV12   PHB
GPU1   NV12   X     PHB
GPU2   PHB   PHB    X
```

### Inside the container: P2P reported as unavailable

```python
>>> import torch
>>> torch.cuda.can_device_access_peer(0, 1)
False
>>> torch.cuda.can_device_access_peer(1, 0)
False
```

`can_device_access_peer` returns `False` for both directions despite 12 active NVLinks. This is the direct cause of all the vLLM warnings — vLLM and NCCL query CUDA P2P capability, get `False`, and disable every NVLink-dependent code path.

### What works and what doesn't

| Feature | Status | Reason |
|---------|--------|--------|
| NCCL communicator init | **Works** | `CUDA_DISABLE_CONTROL=true` fixes the hang (#2641) |
| NCCL AllReduce | **Works** (SHM transport) | NCCL falls back to SHM — slower than NVLink |
| SymmMemCommunicator | **Fails** | `invalid device ordinal` — needs P2P |
| Custom all-reduce | **Fails** | `lacks GPU P2P capability or P2P test failed` |
| FlashInfer MNNVL all-reduce | **Fails** | `invalid device ordinal` — needs NVSwitch multicast |
| AllReduce-norm fusion | **Fails** | Depends on FlashInfer allreduce workspace |
| AllReduce fusion pass | **Disabled** | Cascades from above |

## Root Cause

`CUDA_DISABLE_CONTROL=true` makes HAMi's `libvgpu.so` skip memory allocation control — this fixes the NCCL init hang (where `cuMemCreate` was intercepted and deadlocked). But `libvgpu.so` is **still loaded** via `LD_PRELOAD` and **still intercepts other CUDA calls**, including the P2P capability query (`cuDeviceGetP2PStatus` / `cuDeviceCanAccessPeer`).

The interception returns `False` / `invalid device ordinal` instead of forwarding the query to the real CUDA driver, which would report that NVLink P2P is available. As a result, every component that checks P2P capability before using NVLink sees "not available" and falls back to slower paths.

In other words: `CUDA_DISABLE_CONTROL=true` disables HAMi's *memory control* but does not disable HAMi's *device capability interception*. The `libvgpu.so` library is still in the process address space and still hooks CUDA calls that return device topology and P2P information.

## Impact

- **All-reduce bandwidth**: NCCL uses SHM transport (~23 GB/s) instead of NVLink (~318 GB/s) — potential **10-14x** bandwidth loss for inter-GPU communication.
- **Fusion optimizations disabled**: AllReduce-norm fusion, FlashInfer allreduce, and custom all-reduce are all disabled, adding kernel launch overhead and preventing operator fusion.
- **Tensor parallelism throughput**: Every TP all-reduce operation is slower, directly impacting token generation throughput for TP=2 models.

## Suggested HAMi Improvement

**Fully bypass `libvgpu.so` in whole-GPU mode.** `CUDA_DISABLE_CONTROL=true` currently disables memory allocation control but leaves the library loaded and intercepting capability queries. For whole-GPU allocations (`nvidia.com/gpu: N` without `nvidia.com/gpumem`), there is no reason for `libvgpu.so` to intercept any CUDA call — no memory limit to enforce, no core limit to enforce, no device virtualization needed.

The fix should either:
1. **Unload / skip `libvgpu.so` entirely** when `CUDA_DISABLE_CONTROL=true` is set, so all CUDA calls pass through to the real driver unmodified.
2. **Pass through P2P capability queries** (`cuDeviceGetP2PStatus`, `cuDeviceCanAccessPeer`, `cuDeviceGetNvLinkRemoteDeviceType`, etc.) to the real driver when control is disabled, instead of returning synthetic `False` / error results.

Without this, NVLink is effectively unusable for any workload running under HAMi in whole-GPU mode, regardless of whether the hardware and host VM have NVLink properly configured and active.

**Comment by panpan0000** [comment]
which PR has fixed this ? @ok-claven
